crates/polars-arrow/src/compute/temporal.rs

```diff
@@ -203,7 +203,12 @@ where
                 TimeUnit::Microsecond => timestamp_us_to_datetime,
                 TimeUnit::Nanosecond => timestamp_ns_to_datetime,
             };
-            Ok(unary(array, |x| op(func(x)), data_type))
+            Ok(PrimitiveArray::<O>::from(
+                array
+                    .iter()
+                    .map(|v| v.map(|x| op(func(*x))))
+                    .collect::<Vec<_>>(),
+            ))
```

Comment:
I believe you collect a Vec of options here and then create a PrimitiveArray. This will be a double allocation. You can create the array directly from the iterator.